packages/loopring_v3/contracts/amm/LoopringAmmPool.sol

```diff
@@ -0,0 +1,208 @@
+
+    modifier onlyExchangeOwner()
+    {
+        require(msg.sender == state.exchange.owner(), "UNAUTHORIZED");
+        _;
+    }
+
+    modifier onlyWhenOnline()
+    {
+        require(state.isOnline(), "NOT_ONLINE");
+        _;
+    }
+
+    modifier onlyWhenOffline()
+    {
+        require(!state.isOnline(), "NOT_OFFLINE");
+        _;
+    }
+
+    function isOnline()
+        public
+        view
+        returns (bool)
+    {
+        return state.isOnline();
+    }
+
+    receive() payable external {}
+
+    function init(
+        IExchangeV3        _exchange,
+        uint32             _accountID,
+        address[] calldata _tokens,
+        uint96[]  calldata _weights,
+        uint8              _feeBips,
+        string    calldata _tokenName,
+        string    calldata _tokenSymbol
+        )
+        external
+        nonReentrant
+    {
+        require(
+            bytes(_tokenName).length > 0 && bytes(_tokenSymbol).length > 0,
+            "INVALID_NAME_OR_SYMBOL"
+        );
+        state.name = _tokenName;
+        state.symbol = _tokenSymbol;
+
+        state.setupPool(_exchange, _accountID, _tokens, _weights, _feeBips);
+    }
+
+    // Anyone is able to shut down the pool when requests aren't being processed any more.
+    function shutdown(bytes32 txHash)
+        external
+        payable
+        onlyWhenOnline
+        nonReentrant
+    {
+        state.shutdown(txHash);
+    }
+
+    // Only used to withdraw from the pool when shutdown.
+    // Otherwise LPs should withdraw by doing normal queued exit requests.
+    function withdrawFromPoolWhenShutdown(uint poolAmountIn)
+        external
+        onlyWhenOffline
+        nonReentrant
+    {
+        state.withdrawFromPoolWhenShutdown(poolAmountIn);
+    }
+
+    function deposit(
+        uint96            poolAmount,
+        uint96[] calldata amounts
+        )
+        external
+        payable
+        onlyWhenOnline
+        nonReentrant
+    {
+        state.deposit(poolAmount, amounts);
+        emit Deposit(msg.sender, poolAmount, amounts);
+    }
+
+    function joinPool(
+        uint              minPoolAmountOut,
+        uint96[] calldata maxAmountsIn,
+        bool              fromLayer2,
+        uint              validUntil
+        )
+        external
+        onlyWhenOnline
+        nonReentrant
+    {
+        AmmData.PoolJoin memory join = state.joinPool(
+            minPoolAmountOut,
+            maxAmountsIn,
+            fromLayer2,
+            validUntil
+        );
+        emit PoolJoinRequested(join);
+    }
+
+    function depositAndJoinPool(
+        uint              minPoolAmountOut,
+        uint96[] calldata maxAmountsIn,
+        bool              fromLayer2,
+        uint              validUntil
+        )
+        external
+        onlyWhenOnline
+        nonReentrant
+    {
+        state.deposit(0, maxAmountsIn);
+        state.joinPool(minPoolAmountOut, maxAmountsIn, fromLayer2, validUntil);
```

Comment:
Probably best to move the events `emit Deposit(msg.sender, poolAmount, amounts);` and `emit PoolJoinRequested(join);` to the internal functions so they are always emitted.

Reply:
done